include/boost/math/tools/differential_evolution.hpp

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright Nick Thompson, 2023
+ * Use, modification and distribution are subject to the
+ * Boost Software License, Version 1.0. (See accompanying file
+ * LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+ */
+#ifndef BOOST_MATH_TOOLS_DIFFERENTIAL_EVOLUTION_HPP
+#define BOOST_MATH_TOOLS_DIFFERENTIAL_EVOLUTION_HPP
+#include <algorithm>
+#include <array>
+#include <atomic>
+#include <cmath>
+#include <limits>
+#if __has_include(<omp.h>)
+#include <omp.h>
+#endif
+#include <random>
+#include <stdexcept>
+#include <sstream>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace boost::math::tools {
+
+namespace detail {
+template <class T>
+concept subscriptable = requires(const T &c) { c[0]; };
+}
+// Storn, R., Price, K. (1997). Differential evolution-a simple and efficient heuristic for global optimization over
+// continuous spaces.
+// Journal of global optimization, 11, 341-359.
+// See:
+// https://www.cp.eng.chula.ac.th/~prabhas//teaching/ec/ec2012/storn_price_de.pdf
+template <typename BoundsContainer>
+  requires detail::subscriptable<BoundsContainer>
+class differential_evolution {
+public:
+  using BoundType = BoundsContainer::value_type;
+  using Real = BoundType::value_type;
+  differential_evolution(BoundsContainer bounds, Real F = 0.65, double crossover_ratio = 0.5, size_t NP = 200,
+                         size_t max_generations = 1000, size_t threads = std::thread::hardware_concurrency())
```

Comment:
```suggestion
                         size_t max_generations = 1000, int threads = std::thread::hardware_concurrency())
```

This would warn of implicit unsigned to signed conversion.

Reply:
You're right. I thought it was int but it's actually unsigned: https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency